### PR TITLE
procedures: update admin docs for storage classes, storage strategies and storage size

### PR DESCRIPTION
**Pull request by AObuchow**
## What does this pull request change?
- Update to the storage class documentation:
  - Remove mention of Che Server creating PVC's as this is now handled by DevWorkspace Operator
  - Update procedure for specifying storage class to be up-to-date with Che Operator CRD
- Add additional documentation for configuration of storage
  - Explain storage strategies available, as well as default storage strategy
  - Add procedure for switching storage strategies
  - Explain how storage sizes (persistent volume claim sizes) are configured (i.e. following Kubernetes resource quantity format)
  - Add procedure for changing storage sizes
 

**Some notes:**
- There are some redundancies that I included for clarity, but may be acting as noise:
  - For both procedures in the "Configuring storage strategy" page, there are links to configuring the Che Cluster CR
  - For the code examples, there are duplicate comments because there are duplicate fields

## What issues does this pull request fix or reference?
https://github.com/eclipse/che/issues/21885#issuecomment-1372223130

## Specify the version of the product this pull request applies to

7.58.0 (I believe)

## Pull Request checklist

The author and the reviewers validate the content of this pull request with the following checklist, in addition to the [automated tests](code_review_checklist.adoc).

- Any procedure:
  - [x] Successfully tested.
- Any page or link rename:
  - [ ] The page contains a redirection for the previous URL.
  - Propagate the URL change in:
    - [ ] Dashboard [default branding data](https://github.com/eclipse-che/che-dashboard/blob/main/packages/dashboard-frontend/src/services/bootstrap/branding.constant.ts)
    - [ ] Chectl [constants.ts](https://github.com/che-incubator/chectl/blob/main/src/constants.ts)
- [x] Builds on [Eclipse Che hosted by Red Hat](https://workspaces.openshift.com).
- [x] the *`Validate language on files added or modified`* step reports no vale warnings.

### CODE REVIEW

**Comment by max-cx** [comment]
Thank you for this PR, Andrew. A technical writer will self-assign it to review it toward [RHDEVDOCS-4870](https://issues.redhat.com/browse/RHDEVDOCS-4870).

**Comment by AObuchow** [comment]
@max-cx Sounds good, thank you for the update :)

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (AObuchow) [inline modules/administration-guide/pages/configuring-che-storage-size.adoc]: Not sure, but it might be a bit redundant to say "in Che/DevSpaces" since this is the Che/DevSpaces docs.

File: modules/administration-guide/pages/configuring-che-storage-strategy.adoc
Comment (AObuchow) [inline modules/administration-guide/pages/configuring-che-storage-strategy.adoc]: While working on this, I realized that the documentation on requesting persistent storage for workspaces with a devfile (`requesting-persistent-storage-for-workspaces.adoc`) also should probably be updated to reference the [DevWorkspace Operator storage strategy attribute](https://github.com/devfile/devworkspace-operator/blob/main/docs/additional-configuration.adoc#configuring-persistent-storage-used-for-a-devworkspace)

**Comment by AObuchow** [comment]
I made a separate page for the procedure regarding configuring the PVC/storage size.

**Comment by max-cx** [comment]
@AObuchow, @osslate will review this PR for [RHDEVDOCS-4870](https://issues.redhat.com/browse/RHDEVDOCS-4870).

**Comment by max-cx** [comment]
Picking up the language review as [RHDEVDOCS-5132](https://issues.redhat.com/browse/RHDEVDOCS-5132)

**Comment by AObuchow** [comment]
@max-cx I fixed the issue with the ECA (the commits had the wrong author)

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (max-cx) [reply modules/administration-guide/pages/configuring-che-storage-size.adoc:8]: The `id` and the `=` title have to be the same (except for the lowercase dashed syntax).

Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-size.adoc]: The `id` and the `=` heading have to be the same (except for the lowercase dashed syntax).
I see there's a sibling page with the heading that reads "[Configuring storage classes](https://www.eclipse.org/che/docs/stable/administration-guide/installing-che-using-storage-classes/)".

Also, our convention is for the file name to be the same as the `id`.
Feel free to change the file names, just beware of having to update all the `xref`s to the updated file names in the other files in this PR, esp. in `modules/administration-guide/nav.adoc` and `modules/administration-guide/pages/configuring-storage.adoc`.

For example (and not sure if `size` should be singular or plural here):
```suggestion
[id="configuring-storage-size"]
= Configuring storage size
```

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-size.adoc]: If you update the `id` and `=` title below, the `:description:` (line 2) and `:navtitle:` (line 4) values have to be updated too to be the same as the `=` title.

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-size.adoc]: We use the `link:` variation of AsciiDoc syntax.

We need to use active voice rather than passive voice where possible.
Also, we need to focus on the reader and write to the second person, `you`,  or use the imperative form `[do] [this]`.

For example, assuming the proposed changes are still technically correct:
```suggestion
You can configure the persistent volume claim (PVC) size when using the `per-user` or `per-workspace` storage strategies. You must specify the PVC sizes in the `CheCluster` Custom Resource in the format of a link:https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/[{kubernetes} resource quantity]. For more details on the available storage strategies, see xref:configuring-che-storage-strategy.adoc[].
```

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-size.adoc]: Correct me if I'm wrong, this seems to be a procedure:
```suggestion
:_content-type: PROCEDURE
```

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-size.adoc]: FYI, we have an [online guide about modular docs](https://redhat-documentation.github.io/modular-docs/#con-creating-procedure-modules_writing-mod-docs).

We reserve the `._<Heading>_` syntax for three sections in a procedure module:
`.Prerequisites` (only if any are included, otherwise omitted)
`.Procedure`
`.Additional resources`  (only if any are included, otherwise omitted)

So I suggest removing the `.` and making that line plain text:
```suggestion
Default persistent volume claim sizes:
```

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-size.adoc]: After this point, we need a numbered list.
In AsciiDoc, the numbers can also be added automatically, so you can just start each step as follows (dot, space, sentence) without worrying about putting the correct numbers on the steps:
`. This is a sentence for step 1.`
`. This is a sentence for step 2.`
`. This is a sentence for step 3.`

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-size.adoc]: This sentence appears to summarize the use case and if so, then it belongs in the intro paragraph or two that currently starts on line 10.

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-size.adoc]: WIP

You can move the second sentence into an admonition (NOTE/TIP/IMPORTANT/WARNING) if the info is not one of the steps.

```suggestion
. Set the appropriate `claimSize` field for the desired storage strategy in the Che Cluster Custom Resource:

[NOTE]
====

* You can set this field at installation. See xref:using-chectl-to-configure-the-checluster-custom-resource-during-installation.adoc[].

* You can update this field on the command line. See xref:using-the-cli-to-configure-the-checluster-custom-resource.adoc[].

====
```

File: modules/administration-guide/pages/configuring-che-storage-strategy.adoc
Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-strategy.adoc]: Same comment as for https://github.com/eclipse-che/che-docs/pull/2536/files#r1157449191.

File: modules/administration-guide/pages/configuring-che-storage-strategy.adoc
Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-strategy.adoc]: Same comment as in https://github.com/eclipse-che/che-docs/pull/2536/files#r1157461859.

File: modules/administration-guide/pages/configuring-che-storage-strategy.adoc
Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-strategy.adoc]: Same comment as in https://github.com/eclipse-che/che-docs/pull/2536/files#r1157451521.

**Comment by AObuchow** [comment]
@max-cx thank you so much for the feedback so far :) Will try and address everything soon enough - sorry for the delay on my end

File: modules/administration-guide/pages/configuring-che-storage-strategy.adoc
Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-strategy.adoc]:
```suggestion
Set the `pvcStrategy` field in the Che Cluster Custom Resource to `per-user`, `per-workspace` or `ephemeral`.

[NOTE]
====

* You can set this field at installation. See xref:using-chectl-to-configure-the-checluster-custom-resource-during-installation.adoc[].

* You can update this field on the command line. See xref:using-the-cli-to-configure-the-checluster-custom-resource.adoc[].

====

```

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (max-cx) [reply modules/administration-guide/pages/configuring-che-storage-size.adoc]: Yes, we can avoid mentioning Che/DevSpaces when it's self-explanatory to our readers.

File: modules/administration-guide/pages/installing-che-using-storage-classes.adoc
Comment (max-cx) [inline modules/administration-guide/pages/installing-che-using-storage-classes.adoc]:
```suggestion


```

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-size.adoc]: AsciiDoc offers some advanced options. You can try this if you want:
```suggestion
* {empty}
+
[source,yaml]
----
per-user: 10Gi
----

* {empty}
+
[source,yaml]
----
per-workspace: 5Gi
----
```
BTW, here's the lengthy AsciiDoc documentation, so it might be handy for you to bookmark this link for future reference:
https://docs.asciidoctor.org/asciidoc/latest/

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (max-cx) [inline modules/administration-guide/pages/configuring-che-storage-size.adoc]: Rather than use in-block comments, AsciiDoc allows us to mark lines with numbered callouts and then add legends for the callout numbers under the block:
```suggestion
[source,yaml,subs="+quotes,+attributes"]
----
spec:
  devEnvironments:
    storage:
      pvc:
        pvcStrategy: '__<strategy_name>__'  <1>
        perUserStrategyPvcConfig: <2>
          claimSize: __<size_number>__Gi
        perWorkspaceStrategyPvcConfig:  <3>
          claimSize: __<size_number>__Gi
----
<1> Select the storage strategy: `per-user` or * `per-workspace` or `ephemeral`.
<2> Specify a claim size on the next line or omit the next line to set the default claim size value. The specified claim size is only used when you select this storage strategy.
<3> Specify a claim size on the next line or omit the next line to set the default claim size value. The specified claim size is only used when you select this storage strategy.
```

Please double-check whether the callout numbers in my suggestion above are on the correct lines. If not, then please edit this suggestion and move the callout numbers to the relevant lines.

Is there a `ephemeralStrategyPvcConfig` line? Or does `ephemeral` strategy not need such a line? Either way, please add this info explicitly.

Do these fields support only `Gi`? Do they support other units? If so, please let me know in the comments below.

File: modules/administration-guide/pages/installing-che-using-storage-classes.adoc
Comment (max-cx) [inline modules/administration-guide/pages/installing-che-using-storage-classes.adoc]:
```suggestion
        pvcStrategy: 'per-user' <1>
        perUserStrategyPvcConfig: <2>
          storageClass: 'workspace-storage'
        perWorkspaceStrategyPvcConfig: <3>
          storageClass: 'workspace-storage'        
----
<1> Select the storage strategy: `per-user` or * `per-workspace` or `ephemeral`.
<2> Specify a storage class on the next line or omit the next line to set the default storage class. The specified storage class is only used when you select this storage strategy.
<3> Specify a storage class on the next line or omit the next line to set the default storage class. The specified storage class is only used when you select this storage strategy.
```

Also, I have a question: what are other values, besides 'workspace-storage', that an admin can enter for `storageClass:`? Please try to list those like what I did for `<1>` of the suggestion.

File: modules/administration-guide/pages/configuring-che-storage-strategy.adoc
Comment (max-cx) [reply modules/administration-guide/pages/configuring-che-storage-strategy.adoc]: Please do if you can. Otherwise, please open another issue so that it can be assigned to someone to work on at a later time.

**Review by max-cx — DISMISSED** [review]
@AObuchow, I think this is all from me for now. I might have more suggestions once you've responded to my current comments.

Also, there is a conflict that needs to be resolved. However, you might want to first address my comments, which might make it easier to deal with.

And apologies for the unexpected lack of a preview. I'm hoping it will show up eventually as more changes are made to the PR. If not, then I'll do something about it.

**Comment by AObuchow** [comment]
Hi @max-cx thank you for the great feedback so far. I just rebased my PR cause I was having some issues with the live preview on the dog fooding instance. Hopefully this should fix the issue I was encountering, and I will be starting to address your feedback finally.

Will ping you once it's all addressed (this might take a few sessions of me working on this) :)

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (AObuchow) [reply modules/administration-guide/pages/configuring-che-storage-size.adoc]: For now, I'm going to remove this sentence as it doesn't seem to add any extra information that the first paragraph doesn't already have.

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (AObuchow) [reply modules/administration-guide/pages/configuring-che-storage-size.adoc]: > Please double-check whether the callout numbers in my suggestion above are on the correct lines.

They are correct :) Though, since callout `<2>` and `<3>` are the same, maybe we can get rid of callout `<3>` and reuse callout `<2>` twice? 

> Is there a `ephemeralStrategyPvcConfig` line? Or does `ephemeral` strategy not need such a line? Either way, please add this info explicitly.

There is no `ephemeralStrategyPvcConfig` (as ephemeral implies non-persistent storage, i.e. all stored files will be deleted when the workspace is stopped). I added a note regarding this in the first callout:

```asciidoc
<1> Select the storage strategy: `per-user` or * `per-workspace` or `ephemeral`. Note: the `ephemeral` storage strategy does not use persistent storage, therefore you cannot configure its storage size or other PVC-related attributes.
```

Not sure if this info should be a part of the callout, or somewhere else. I think having it as part of the callout makes sense, however

> 
> Do these fields support only `Gi`? Do they support other units? If so, please let me know in the comments below.

Other units are supported as they are [Kubernetes resource quantities](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/). I find the documentation Kubernetes provides to be a little confusing, however, so I consulted some other explanations of the units available for storage sizes. 

[This page](https://medium.com/swlh/understanding-kubernetes-resource-cpu-and-memory-units-30284b3cc866) gives a nice breakdown, see the "Memory Units" section.

## To summarize the storage size units:
Storage sizes are defined in bytes. There are 4 different ways of specifying the number of bytes for the storage size.

### Explicit Integer
You can define the size as an explicit integer to specify the exact number of bytes, e.g. `128974848`

### Exponent
You can use exponent notation using `e`, for example: `129e6`, this results in some rounding and is an approximation of the value you enter. 

### Fixed-point suffix

The following fixed-point suffixes are available, and are similar to storage units we typically use on consumer hardware (e.g. hardrives, flash drives, etc): `E, P, T, G, M, K`

- E = Exabyte. 1E = 1,000,000,000,000,000,000 bytes
- P = Petabyte. 1P = 1,000,000,000,000,000 bytes
- T = Terabyte. 1T = 1,000,000,000,000 bytes
- G = Gigabyte. 1G = 1,000,000,000 bytes
- M = Megabyte. 1M = 1,000,000 bytes
- K = Kilobyte. 1K = 1,000 bytes

### Binary, Power-of-Two Suffix

The following power-of-two suffixes are available:  `Ei, Pi, Ti, Gi, Mi, Ki` (it seems you can also add a trailing B, e.g. `EiB, PiB, TiB, GiB, MiB, KiB`, though this detail can probably be ignored for simplicity)

- Ei = EiB = Exbibyte. 1Ei = 2⁶⁰ = 1,152,921,504,606,846,976 bytes
- Pi = PiB = Pebibyte. 1Pi = 2⁵⁰ = 1,125,899,906,842,624 bytes
- Ti = TiB = Tebibyte. 1Ti = 2⁴⁰ = 1,099,511,627,776 bytes
- Gi = GiB = Gibibyte. 1Gi = 2³⁰ = 1,073,741,824 bytes
- Mi = MiB = Mebibyte. 1Mi = 2²⁰ = 1,048,576 bytes
- Ki = KiB = Kibibyte. 1Ki = 2¹⁰ = 1,024 bytes


### Thoughts on storage size units:
- In practice, I don't think users will use explicit integers or exponent notation for specifying sizes. The most commonly used size unit that I've seen is binary power-of-two suffix notation, though fixed-point suffix notation seems reasonable to use as it is more commonly seen on consumer storage hardware. 
- Since the [official Kubernetes documentation](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/) on resource quantity is a bit confusing IMO, it may be worth having the above information explicitly stated in the Che docs? Either on the configuring storage size page, or as a separate docs page.

File: modules/administration-guide/pages/installing-che-using-storage-classes.adoc
Comment (AObuchow) [reply modules/administration-guide/pages/installing-che-using-storage-classes.adoc]: I still haven't updated the page on storage classes (going to take a break for today) but I wanted to try and answer this question already.

The `storageClass` field can be set to the name of a [Kubernetes storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/) that exists on the cluster where Che is installed. IIRC, storage class names follow the [following naming convention](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names), though this detail is probably outside the scope of Che docs. 

I can figure out how to explain this detail in the callout the next time I work on this PR.

**Comment by AObuchow** [comment]
@max-cx I've added an extra PR to address some of the feedback you've given, though I still need to do some more work on this PR and finish addressing your feedback. You're welcome to wait until I have finished addressing everything before getting back to me, since this is currently a work-in-progress :P 

And sorry about the ECA, my git email/name is not configured correctly on the dog fooding instance where I worked on this PR. ~~I will fix that before letting you know that the PR is ready for a second review~~ (unless there's urgency for me to fix this).

Edit: Fixing the ECA will probably require me to re-author the commits. If there's no urgency for me to do this, I'll leave it for when the PR is ready to merge as we'll likely want to squash and reorganize the commit log.

File: modules/administration-guide/pages/configuring-che-storage-size.adoc
Comment (AObuchow) [reply modules/administration-guide/pages/configuring-che-storage-size.adoc]: As it currently stands, the pages on configuring the storage strategy and configuring the storage size only have a single step. Should I still be using the `.Procedure` heading?

**Comment by AObuchow** [comment]
@max-cx I've finally done the rest of the updates to this PR based on your current feedback. When you have some time, let me know which further changes are required. For your convenience, I'd be happy to any details on a call if you'd prefer :)

File: modules/administration-guide/pages/configuring-che-storage-strategy.adoc
Comment (AObuchow) [reply modules/administration-guide/pages/configuring-che-storage-strategy.adoc]: I've created https://github.com/eclipse/che/issues/22316 to track this issue and determine if it is valid, as it may not be something Che users need to know about.

**Comment by ibuziuk** [comment]
@AObuchow could you please rebase against main?

**Comment by AObuchow** [comment]
> @AObuchow could you please rebase against main?

Done :+1:

**Comment by ibuziuk** [comment]
@AObuchow looks like still some checks are failing

**Review by ibuziuk — APPROVED** [review]

**Comment by ibuziuk** [comment]
Approved
@deerskindoll @max-cx could we merge it once all the checks are green? PR was opened more than 8 months ago

**Comment by deerskindoll** [comment]
> Approved @deerskindoll @max-cx could we merge it once all the checks are green? PR was opened more than 8 months ago

Sure! How long do you think it will take to clear out the eclipsefdn/eca check? Update: I see, the merge is actually blocked by a change request. @AObuchow can you please resolve the change request?

**Review by deerskindoll — APPROVED** [review]
lgtm

**Comment by AObuchow** [comment]
> > Approved @deerskindoll @max-cx could we merge it once all the checks are green? PR was opened more than 8 months ago
> 
> Sure! How long do you think it will take to clear out the eclipsefdn/eca check? Update: I see, the merge is actually blocked by a change request. @AObuchow can you please resolve the change request?

Thanks for checking this out @deerskindoll :) I believe I had already addressed the changes requested. I've squashed my commits, though [`0cc5c7e` (#2536)](https://github.com/eclipse-che/che-docs/pull/2536/commits/0cc5c7ee0da3722cb05da87d9b6db261e3b31d85) and [`2e93b12` (#2536)](https://github.com/eclipse-che/che-docs/pull/2536/commits/2e93b1258e1661e50efe54ba5f1985d428bcfd4d) could both be squashed into a single commit potentially. Let me know if there's anything missing on my end to get this merged.

File: modules/administration-guide/pages/configuring-storage-sizes.adoc
Comment (deerskindoll) [inline modules/administration-guide/pages/configuring-storage-sizes.adoc]:
```suggestion
You can configure the persistent volume claim (PVC) size using the `per-user` or `per-workspace` storage strategies. You must specify the PVC sizes in the `CheCluster` Custom Resource in the format of a link:https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/[{kubernetes} resource quantity]. For more details on the available storage strategies, see xref:configuring-the-storage-strategy.adoc[this page].
```

**Review by deerskindoll — COMMENTED** [review]
added missing xref description

**Pull request merged**
